File: src/utils/history.ts
// Reading history, persisted as ~/.config/manga-cli/history.json.

import { mkdir } from "node:fs/promises";
import { HISTORY_FILE, CONFIG_DIR } from "./paths.ts";
import type { SourceId } from "../api/types.ts";

export interface HistoryEntry {
  id: string;
  title: string;
  source?: SourceId;
  coverUrl?: string;
  lastChapterId: string;
  lastChapterNumber: number;
  lastChapterTitle: string;
  lastChapterIndex: number;
  lastPage: number;
  totalChapters: number;
  lastReadAt: string; // ISO 8601
}

interface HistoryFile {
  history: HistoryEntry[];
}

const MAX_ENTRIES = 200;

function titleKey(t: string): string {
  return (t ?? "").trim().toLowerCase();
}

export async function loadHistory(): Promise<HistoryEntry[]> {
  try {
    const file = Bun.file(HISTORY_FILE);
    if (!(await file.exists())) return [];
    const data = (await file.json()) as Partial<HistoryFile>;
    const all = data.history ?? [];
    // Collapse duplicate titles (the same series read from different sources)
    // into one entry — entries are most-recent-first, so keep the first seen.
    const seen = new Set<string>();
    return all.filter((h) => {
      const k = titleKey(h.title);
      if (seen.has(k)) return false;
      seen.add(k);
      return true;
    });
  } catch {
    return [];
  }
}

/** Upsert an entry by id *or title* and move it to the front (most-recent-first). */
export async function recordHistory(entry: HistoryEntry): Promise<void> {
  const k = titleKey(entry.title);
  const history = (await loadHistory()).filter((h) => h.id !== entry.id && titleKey(h.title) !== k);
  history.unshift(entry);
  const trimmed = history.slice(0, MAX_ENTRIES);
  try {
    await mkdir(CONFIG_DIR, { recursive: true });
    await Bun.write(HISTORY_FILE, JSON.stringify({ history: trimmed }, null, 2));
  } catch {
    // History is best-effort.
  }
}

export async function getHistoryEntry(id: string): Promise<HistoryEntry | undefined> {
  return (await loadHistory()).find((h) => h.id === id);
}

export async function mostRecent(): Promise<HistoryEntry | undefined> {
  return (await loadHistory())[0];
}

File: go-rewrite/internal/util/history.go
// Reading history — same ~/.config/manga-cli/history.json as the TS version,
// including the dedup-by-title behavior (one entry per series across sources).

package util

import (
	"encoding/json"
	"os"
	"strings"
)

type HistoryEntry struct {
	ID                string  `json:"id"`
	Title             string  `json:"title"`
	Source            string  `json:"source,omitempty"`
	CoverURL          string  `json:"coverUrl,omitempty"`
	LastChapterID     string  `json:"lastChapterId"`
	LastChapterNumber float64 `json:"lastChapterNumber"`
	LastChapterTitle  string  `json:"lastChapterTitle"`
	LastChapterIndex  int     `json:"lastChapterIndex"`
	LastPage          int     `json:"lastPage"`
	TotalChapters     int     `json:"totalChapters"`
	LastReadAt        string  `json:"lastReadAt"` // ISO 8601
}

type historyFile struct {
	History []HistoryEntry `json:"history"`
}

const maxHistory = 200

func titleKey(t string) string { return strings.ToLower(strings.TrimSpace(t)) }

// LoadHistory returns entries most-recent-first, duplicates collapsed by title.
func LoadHistory() []HistoryEntry {
	raw, err := os.ReadFile(HistoryFile)
	if err != nil {
		return nil
	}
	var f historyFile
	if json.Unmarshal(raw, &f) != nil {
		return nil
	}
	seen := map[string]bool{}
	var out []HistoryEntry
	for _, h := range f.History {
		k := titleKey(h.Title)
		if seen[k] {
			continue
		}
		seen[k] = true
		out = append(out, h)
	}
	return out
}

// RecordHistory upserts by id or title and moves the entry to the front.
func RecordHistory(e HistoryEntry) {
	k := titleKey(e.Title)
	all := LoadHistory()
	out := []HistoryEntry{e}
	for _, h := range all {
		if h.ID == e.ID || titleKey(h.Title) == k {
			continue
		}
		out = append(out, h)
	}
	if len(out) > maxHistory {
		out = out[:maxHistory]
	}
	blob, err := json.MarshalIndent(historyFile{History: out}, "", "  ")
	if err != nil {
		return
	}
	if os.MkdirAll(ConfigDir, 0o755) != nil {
		return
	}
	_ = os.WriteFile(HistoryFile, blob, 0o644)
}

// MostRecent returns the newest entry, or nil.
func MostRecent() *HistoryEntry {
	h := LoadHistory()
	if len(h) == 0 {
		return nil
	}
	return &h[0]
}
